File: src/main/java/main/commands/prefixcommands/audio/SkipCommand.java
package main.commands.prefixcommands.audio;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import main.audiohandlers.RobertifyAudioManager;
import main.commands.prefixcommands.CommandContext;
import main.commands.prefixcommands.ICommand;
import main.constants.BotConstants;
import main.constants.Permission;
import main.constants.Toggles;
import main.exceptions.AutoPlayException;
import main.utils.GeneralUtils;
import main.utils.RobertifyEmbedUtils;
import main.utils.json.requestchannel.RequestChannelConfig;
import main.utils.json.logs.LogType;
import main.utils.json.logs.LogUtils;
import main.utils.json.toggles.TogglesConfig;
import main.utils.locale.RobertifyLocaleMessage;
import net.dv8tion.jda.annotations.ForRemoval;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.entities.channel.middleman.GuildMessageChannel;
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import net.dv8tion.jda.api.interactions.components.buttons.Button;
import net.dv8tion.jda.api.interactions.components.buttons.ButtonStyle;
import net.dv8tion.jda.internal.utils.tuple.Pair;
import org.jetbrains.annotations.NotNull;

import javax.script.ScriptException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Stack;

@Deprecated @ForRemoval
public class SkipCommand extends ListenerAdapter implements ICommand {
    private final static HashMap<Long, Integer> voteSkips = new HashMap<>();
    private final static HashMap<Long, ArrayList<Long>> voters = new HashMap<>();
    private final static HashMap<Long, Pair<Long, Long>> voteSkipMessages = new HashMap<>();
    private final static HashMap<Long, Long> voteSkipStarters = new HashMap<>();

    @Override
    public void handle(CommandContext ctx) throws ScriptException {
        final Message msg = ctx.getMessage();
        final Member self = ctx.getSelfMember();
        final GuildVoiceState selfVoiceState = self.getVoiceState();
        final Member member = ctx.getMember();
        final GuildVoiceState memberVoiceState = member.getVoiceState();
        final Guild guild = ctx.getGuild();
        final var togglesConfig = new TogglesConfig(guild);

        if (!togglesConfig.isDJToggleSet(this)) {
            msg.replyEmbeds(handleSkip(selfVoiceState, memberVoiceState)).queue();
            return;
        }

        if (togglesConfig.getDJToggle(this)) {
            if (togglesConfig.getToggle(Toggles.VOTE_SKIPS)) {
                if (!GeneralUtils.hasPerms(guild, member, Permission.ROBERTIFY_DJ)) {
                    if (selfVoiceState.inAudioChannel()) {
                        msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.VOICE_CHANNEL_NEEDED).build())
                                .queue();
                        return;
                    }

                    if (selfVoiceState.getChannel().getMembers().size() != 2) {
                        MessageEmbed embed = handleVoteSkip(ctx.getChannel(), selfVoiceState, memberVoiceState);
                        if (embed != null)
                            msg.replyEmbeds(handleVoteSkip(ctx.getChannel(), selfVoiceState, memberVoiceState)).queue();
                    } else msg.replyEmbeds(handleSkip(selfVoiceState, memberVoiceState)).queue();
                } else msg.replyEmbeds(handleSkip(selfVoiceState, memberVoiceState)).queue();
            } else  {
                if (GeneralUtils.hasPerms(guild, member, Permission.ROBERTIFY_DJ)) {
                    msg.replyEmbeds(handleSkip(selfVoiceState, memberVoiceState)).queue();
                } else msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, BotConstants.getInsufficientPermsMessage(guild, Permission.ROBERTIFY_DJ)).build())
                        .queue();
            }
        } else msg.replyEmbeds(handleSkip(selfVoiceState, memberVoiceState)).queue();
    }

    public MessageEmbed handleSkip(GuildVoiceState selfVoiceState, GuildVoiceState memberVoiceState) {
        final var guild = selfVoiceState.getGuild();

        if (checks(selfVoiceState, memberVoiceState) != null)
            return checks(selfVoiceState, memberVoiceState);

        final var musicManager = RobertifyAudioManager.getInstance().getMusicManager(guild);
        final var audioPlayer = musicManager.getPlayer();

        if (audioPlayer.getPlayingTrack() == null)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SkipMessages.NOTHING_TO_SKIP).build();

        skip(guild);

        new LogUtils(guild).sendLog(LogType.TRACK_SKIP, RobertifyLocaleMessage.SkipMessages.SKIPPED_LOG, Pair.of("{user}", memberVoiceState.getMember().getAsMention()));
        return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SkipMessages.SKIPPED).build();
    }

    public MessageEmbed handleVoteSkip(GuildMessageChannel channel, GuildVoiceState selfVoiceState, GuildVoiceState memberVoiceState) {
        final var guild = selfVoiceState.getGuild();

        if (checks(selfVoiceState, memberVoiceState) != null)
            return checks(selfVoiceState, memberVoiceState);

        final var musicManager = RobertifyAudioManager.getInstance().getMusicManager(guild);
        final var audioPlayer = musicManager.getPlayer();

        if (audioPlayer.getPlayingTrack() == null)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SkipMessages.NOTHING_TO_SKIP).build();

        final int neededVotes = getNeededVotes(guild);
        channel.sendMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild,
                        RobertifyLocaleMessage.SkipMessages.VOTE_SKIP_STARTED_EMBED,
                        Pair.of("{user}", memberVoiceState.getMember().getAsMention()),
                        Pair.of("{neededVotes}", String.valueOf(neededVotes))).build()
                )
                .setActionRow(
                        Button.of(ButtonStyle.SUCCESS, "voteskip:upvote:" + guild.getId(), "Vote"),
                        Button.of(ButtonStyle.DANGER, "voteskip:cancel:" + guild.getId(), "Cancel")
                ).queue(success -> {
                    new LogUtils(guild).sendLog(LogType.TRACK_VOTE_SKIP, RobertifyLocaleMessage.SkipMessages.VOTE_SKIP_STARTED_LOG, Pair.of("{user}", memberVoiceState.getMember().getAsMention()));
                    voteSkips.put(guild.getIdLong(), 1);
                    voteSkipMessages.put(guild.getIdLong(), Pair.of(channel.getIdLong(), success.getIdLong()) );

                    ArrayList<Long> list = new ArrayList<>();
                    list.add(memberVoiceState.getIdLong());
                    voters.put(guild.getIdLong(), list);

                    voteSkipStarters.put(guild.getIdLong(), memberVoiceState.getMember().getIdLong());
                });
        return null;
    }

    private void skip(Guild guild) {
        final var musicManager = RobertifyAudioManager.getInstance().getMusicManager(guild);
        final var audioPlayer = musicManager.getPlayer();
        final var scheduler = musicManager.getScheduler();

<<<<<<< HEAD
        final var playingTrack = audioPlayer.getPlayingTrack();
        final var pastQueue = musicManager.getScheduler().getPastQueue();
        pastQueue.push(audioPlayer.getPlayingTrack());
=======
        AudioTrack playingTrack = audioPlayer.getPlayingTrack();
        HashMap<Long, Stack<AudioTrack>> pastQueue = musicManager.getScheduler().getPastQueue();
        if (!pastQueue.containsKey(guild.getIdLong()))
            pastQueue.put(guild.getIdLong(), new Stack<>());
        pastQueue.get(guild.getIdLong()).push(audioPlayer.getPlayingTrack());

        if (scheduler.isRepeating())
            scheduler.setRepeating(false);
>>>>>>> 1f173d748606af2f96273701cdce66240218a441

        if (scheduler.isRepeating())
            scheduler.setRepeating(false);

        try {
            musicManager.getScheduler().nextTrack(playingTrack, true, audioPlayer.getTrackPosition());
        } catch (AutoPlayException e) {
            scheduler.getAnnouncementChannel().sendMessageEmbeds(RobertifyEmbedUtils.embedMessageWithTitle(guild,
                    "AutoPlay","Could not auto-play because tracks from `"+ playingTrack.getSourceManager().getSourceName()+"` aren't supported!").build())
                    .queue();
            scheduler.scheduleDisconnect(true);
        }

        if (new RequestChannelConfig(guild).isChannelSet())
            new RequestChannelConfig(guild).updateMessage();

        clearVoteSkipInfo(guild);
    }

    public static void clearVoteSkipInfo(Guild guild) {
        voters.remove(guild.getIdLong());
        voteSkips.remove(guild.getIdLong());
        voteSkipStarters.remove(guild.getIdLong());

        if (voteSkipMessages.get(guild.getIdLong()) != null) {
            Pair<Long, Long> pair = voteSkipMessages.get(guild.getIdLong());
            guild.getTextChannelById(pair.getLeft()).retrieveMessageById(pair.getRight()).complete()
                    .editMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SkipMessages.SKIPPED)
                            .build())
                    .queue();
        }

        voteSkipMessages.remove(guild.getIdLong());
    }

    private MessageEmbed checks(GuildVoiceState selfVoiceState, GuildVoiceState memberVoiceState) {
        final var guild = selfVoiceState.getGuild();

        if (!selfVoiceState.inAudioChannel())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.NOTHING_PLAYING).build();

        if (!memberVoiceState.inAudioChannel())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.USER_VOICE_CHANNEL_NEEDED).build();

        if (!memberVoiceState.getChannel().equals(selfVoiceState.getChannel()))
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL_LOC, Pair.of("{channel}", selfVoiceState.getChannel().getAsMention())).build();

        return null;
    }

    public boolean voteSkipIsActive(Guild guild) {
        return voteSkips.containsKey(guild.getIdLong());
    }

    private void incrementVoteSkip(Guild guild) {
        if (!voteSkipIsActive(guild))
            throw new IllegalStateException("Can't increment vote skip since guild doesn't have any active vote skips.");

        voteSkips.put(guild.getIdLong(), voteSkips.get(guild.getIdLong()) + 1);

        final int neededVotes = getNeededVotes(guild);
        if (voteSkips.get(guild.getIdLong()) == neededVotes)
            doVoteSkip(guild);
    }

    private void doVoteSkip(Guild guild) {
        Pair<Long, Long> pair = voteSkipMessages.get(guild.getIdLong());
        Message message = guild.getTextChannelById(pair.getLeft())
                .retrieveMessageById(pair.getRight()).complete();

        voteSkipMessages.remove(guild.getIdLong());

        AudioTrackInfo info = RobertifyAudioManager.getInstance()
                .getMusicManager(guild)
                .getPlayer()
                .getPlayingTrack()
                .getInfo();

        skip(guild);
        message.editMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SkipMessages.VOTE_SKIPPED,
                    Pair.of("{title}", info.title),
                    Pair.of("{author}", info.author)
                ).build())
                .setComponents()
                .queue();

        new LogUtils(guild).sendLog(LogType.TRACK_SKIP, RobertifyLocaleMessage.SkipMessages.VOTE_SKIPPED_LOG,
                Pair.of("{title}", info.title),
                Pair.of("{author}", info.author)
        );
    }

    private void decrementVoteSkip(Guild guild) {
        if (!voteSkipIsActive(guild))
            throw new IllegalStateException("Can't decrement vote skip since guild doesn't have any active vote skips.");

        if (voteSkips.get(guild.getIdLong()) == 0)
            throw new IllegalStateException("Can't decrement vote skip past zero!");

        voteSkips.put(guild.getIdLong(), voteSkips.get(guild.getIdLong()) - 1);
    }

    private void updateVoteSkipMessage(Guild guild) {
        if (!voteSkipIsActive(guild))
            return;

        Pair<Long, Long> pair = voteSkipMessages.get(guild.getIdLong());
        Message message = guild.getTextChannelById(pair.getLeft()).retrieveMessageById(pair.getRight()).complete();

        final int neededVotes = getNeededVotes(guild);
        message.editMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SkipMessages.VOTE_SKIP_STARTED_EMBED,
                    Pair.of("{user}", GeneralUtils.toMention(guild, voteSkipStarters.get(guild.getIdLong()), GeneralUtils.Mentioner.USER)),
                    Pair.of("{neededVotes}", String.valueOf(neededVotes))
                ).build()).queue();
    }

    private int getNeededVotes(Guild guild) {
        final int size = guild.getSelfMember().getVoiceState().getChannel().getMembers().size();
        return (int)Math.ceil(size * (50/100.0));
    }

    @Override
    public void onButtonInteraction(@NotNull ButtonInteractionEvent event) {
        if (!event.getButton().getId().startsWith("voteskip:"))
            return;

        String[] split = event.getButton().getId().split(":");
        final String voteType = split[1];
        final String guildID = split[2];

        Guild guild = event.getGuild();
        if (!guild.getId().equals(guildID))
            return;

        GuildVoiceState voiceState = event.getGuild().getSelfMember().getVoiceState();
        GuildVoiceState memberState = event.getMember().getVoiceState();

        if (!voiceState.inAudioChannel()) {
            event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.BUTTON_NO_LONGER_VALID).build())
                    .setEphemeral(true)
                    .queue();
            return;
        }

        if (!memberState.inAudioChannel()) {
            event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL_BUTTON).build())
                    .setEphemeral(true)
                    .queue();
            return;
        }

        if (!voiceState.getChannel().equals(memberState.getChannel())) {
            event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL_BUTTON).build())
                    .setEphemeral(true)
                    .queue();
            return;
        }

        User user = event.getUser();
        switch (voteType) {
            case "upvote" -> {
                ArrayList<Long> list = voters.get(guild.getIdLong());
                if (list.contains(user.getIdLong())) {
                    list.remove(user.getIdLong());
                    voters.put(guild.getIdLong(), list);
                    decrementVoteSkip(guild);
                    event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SkipMessages.SKIP_VOTE_REMOVED).build())
                            .setEphemeral(true)
                            .queue();
                    updateVoteSkipMessage(guild);
                } else {
                    if (GeneralUtils.hasPerms(guild, event.getMember(), Permission.ROBERTIFY_DJ)) {
                        doVoteSkip(guild);
                        event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SkipMessages.DJ_SKIPPED).build())
                                .setEphemeral(true)
                                .queue();
                        return;
                    }

                    list.add(user.getIdLong());
                    voters.put(guild.getIdLong(), list);
                    incrementVoteSkip(guild);
                    event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.SkipMessages.SKIP_VOTE_ADDED).build())
                            .setEphemeral(true)
                            .queue();
                    updateVoteSkipMessage(guild);
                }
            }
            case "cancel" -> {
                if (voteSkipStarters.get(guild.getIdLong()) != user.getIdLong()) {
                    event.replyEmbeds(RobertifyEmbedUtils.embedMessage(
                            guild,
                            RobertifyLocaleMessage.GeneralMessages.NO_PERMS_BUTTON
                    ).build())
                            .setEphemeral(true)
                            .queue();
                    return;
                }

                Pair<Long, Long> pair = voteSkipMessages.get(guild.getIdLong());
                Message message = guild.getTextChannelById(pair.getLeft())
                        .retrieveMessageById(pair.getRight()).complete();

                AudioTrackInfo info = RobertifyAudioManager.getInstance()
                        .getMusicManager(guild)
                        .getPlayer()
                        .getPlayingTrack()
                        .getInfo();

                voteSkipMessages.remove(guild.getIdLong());
                clearVoteSkipInfo(guild);
                message.editMessageEmbeds(RobertifyEmbedUtils.embedMessage(
                        guild,
                        RobertifyLocaleMessage.SkipMessages.VOTE_SKIP_CANCELLED,
                        Pair.of("{title}", info.title),
                        Pair.of("{author}", info.author)
                        ).build()
                ).queue();
            }
        }
    }

    @Override
    public String getName() {
        return "skip";
    }

    @Override
    public String getHelp(String prefix) {
        return "Aliases: `"+ GeneralUtils.listToString(getAliases()) +"`\n" +
                "Skips a track";
    }

    @Override
    public List<String> getAliases() {
        return List.of("next");
    }
}

File: src/main/kotlin/main/commands/slashcommands/audio/SkipCommand.kt
package main.commands.slashcommands.audio

import dev.minn.jda.ktx.interactions.components.danger
import dev.minn.jda.ktx.interactions.components.success
import dev.minn.jda.ktx.util.SLF4J
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import main.audiohandlers.GuildMusicManager
import main.audiohandlers.RobertifyAudioManager
import main.audiohandlers.utils.author
import main.audiohandlers.utils.title
import main.commands.slashcommands.SlashCommandManager.getRequiredOption
import main.constants.RobertifyPermission
import main.utils.GeneralUtils
import main.utils.RobertifyEmbedUtils
import main.utils.RobertifyEmbedUtils.Companion.editEmbed
import main.utils.RobertifyEmbedUtils.Companion.replyEmbed
import main.utils.RobertifyEmbedUtils.Companion.sendEmbed
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.Command
import main.utils.component.interactions.slashcommand.models.CommandOption
import main.utils.json.logs.LogType
import main.utils.json.logs.LogUtilsKt
import main.utils.json.requestchannel.RequestChannelConfig
import main.utils.locale.messages.GeneralMessages
import main.utils.locale.messages.SkipMessages
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.GuildVoiceState
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.entities.channel.middleman.GuildMessageChannel
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent
import net.dv8tion.jda.api.interactions.commands.OptionType
import kotlin.math.ceil

class SkipCommand : AbstractSlashCommand(
    Command(
        name = "skip",
        description = "Skip the song currently being played.",
        options = listOf(
            CommandOption(
                type = OptionType.INTEGER,
                name = "to",
                description = "The ID of the track to skip to",
                required = false
            )
        )
    )
) {

    companion object {
        private val logger by SLF4J
    }

    override suspend fun handle(event: SlashCommandInteractionEvent) {
        event.deferReply().queue()

        val guild = event.guild!!
        val selfVoiceState = guild.selfMember.voiceState!!
        val memberVoiceState = event.member!!.voiceState!!

        if (!musicCommandDJCheck(event)) {
            if (!selfVoiceState.inAudioChannel()) {
                event.hook.sendEmbed(guild) {
                    embed(GeneralMessages.VOICE_CHANNEL_NEEDED)
                }.setEphemeral(true)
                    .queue()
                return
            }

            if (selfVoiceState.channel!!.members.size > 2) {
                val embed = handleVoteSkip(event.channel.asGuildMessageChannel(), selfVoiceState, memberVoiceState)
                if (embed != null) {
                    event.hook.sendMessageEmbeds(embed)
                        .queue()
                } else {
                    event.hook.sendEmbed(guild) {
                        embed(SkipMessages.VOTE_SKIP_STARTED)
                    }.queue()
                }
                return
            }
        }

        if (event.options.isEmpty()) {
            event.hook.sendMessageEmbeds(handleSkip(selfVoiceState, memberVoiceState))
                .queue()
        } else {
            val musicManager = RobertifyAudioManager[guild]
            val tracksToSkip = event.getRequiredOption("to").asLong.toInt()
            event.hook.sendMessageEmbeds(handleSkip(event.user, musicManager, tracksToSkip))
                .queue()
        }
    }

    fun handleSkip(selfVoiceState: GuildVoiceState, memberVoiceState: GuildVoiceState): MessageEmbed {
        val guild = selfVoiceState.guild

        val checksEmbed = audioChannelChecks(selfVoiceState, memberVoiceState)
        if (checksEmbed != null)
            return checksEmbed

        val musicManager = RobertifyAudioManager[guild]
        val player = musicManager.player

        if (player.playingTrack == null)
            return RobertifyEmbedUtils.embedMessage(guild, SkipMessages.NOTHING_TO_SKIP)
                .build()

        skip(guild)

        LogUtilsKt(guild).sendLog(
            LogType.TRACK_SKIP, SkipMessages.SKIPPED_LOG,
            Pair("{user}", memberVoiceState.member.asMention)
        )
        return RobertifyEmbedUtils.embedMessage(guild, SkipMessages.SKIPPED).build()
    }

    fun handleSkip(skipper: User, musicManager: GuildMusicManager, id: Int): MessageEmbed {
        val scheduler = musicManager.scheduler
        val queueHandler = scheduler.queueHandler

        if (id > queueHandler.size || id <= 0)
            return RobertifyEmbedUtils.embedMessage(
                musicManager.guild,
                GeneralMessages.INVALID_ARGS
            ).build()

        val player = musicManager.player
        val playingTrack = player.playingTrack!!
        val guild = musicManager.guild
        val currentQueue = queueHandler.contents.toMutableList()
        val songsToRemove = currentQueue.subList(0, id - 1)

        queueHandler.removeAll(songsToRemove)
        queueHandler.pushPastTrack(playingTrack)
        scheduler.nextTrack(playingTrack, true, playingTrack.position)

        RequestChannelConfig(guild).updateMessage()
        SkipCommand().clearVoteSkipInfo(guild)
        return RobertifyEmbedUtils.embedMessage(musicManager.guild, "Skipped to **track #$id**!").build()
    }

    fun handleVoteSkip(
        channel: GuildMessageChannel,
        selfVoiceState: GuildVoiceState,
        memberVoiceState: GuildVoiceState
    ): MessageEmbed? {
        val guild = selfVoiceState.guild

        val checksEmbed = audioChannelChecks(selfVoiceState, memberVoiceState)
        if (checksEmbed != null)
            return checksEmbed

        val musicManager = RobertifyAudioManager[guild]
        val player = musicManager.player

        if (player.playingTrack == null)
            return RobertifyEmbedUtils.embedMessage(guild, SkipMessages.NOTHING_TO_SKIP)
                .build()

        val neededVotes = getNeededVotes(guild)
        channel.sendEmbed(guild) {
            embed(
                SkipMessages.VOTE_SKIP_STARTED_EMBED,
                Pair("{user}", memberVoiceState.member.asMention),
                Pair("{neededVotes}", neededVotes.toString())
            )
        }
            .setActionRow(
                success(
                    id = "voteskip:upvote:${guild.id}",
                    label = "Vote"
                ),
                danger(
                    id = "voteskip:cancel:${guild.id}"
                )
            )
            .queue { msg ->
                val starter = memberVoiceState.member

                LogUtilsKt(guild).sendLog(
                    LogType.TRACK_VOTE_SKIP, SkipMessages.VOTE_SKIP_STARTED_LOG,
                    Pair("{user}", starter.asMention)
                )

                val voteSkipManager = musicManager.voteSkipManager
                voteSkipManager.startedBy = starter.idLong
                voteSkipManager.voteSkipCount = 1
                voteSkipManager.addVoter(starter.idLong)
                voteSkipManager.voteSkipMessage = Pair(channel.idLong, msg.idLong)
            }
        return null
    }

    private fun skip(guild: Guild) {
        val musicManager = RobertifyAudioManager[guild]
        val audioPlayer = musicManager.player
        val scheduler = musicManager.scheduler
        val playingTrack = audioPlayer.playingTrack
        val queueHandler = scheduler.queueHandler

        if (playingTrack == null) {
            logger.warn("Attempting to skip a null track in ${guild.name}!")
            return
        }

        queueHandler.pushPastTrack(playingTrack)

        if (queueHandler.trackRepeating)
            queueHandler.trackRepeating = false

        scheduler.nextTrack(playingTrack, true, audioPlayer.trackPosition)

        RequestChannelConfig(guild).updateMessage()
        clearVoteSkipInfo(guild)
    }

    private fun doVoteSkip(guild: Guild) {
        val voteSkipManager = RobertifyAudioManager[guild].voteSkipManager
        if (!voteSkipManager.isVoteSkipActive)
            throw IllegalStateException("Can't do a vote skip when there's none active!")
        guild.getTextChannelById(voteSkipManager.voteSkipMessage!!.first)
            ?.retrieveMessageById(voteSkipManager.voteSkipMessage!!.second)
            ?.submit()
            ?.whenComplete { msg, err ->
                if (err != null) {
                    logger.error("Unexpected error", err)
                    return@whenComplete
                }

                voteSkipManager.voteSkipMessage = null
                val track = RobertifyAudioManager[guild]
                    .player
                    .playingTrack!!

                runBlocking {
                    launch { skip(guild) }
                }

                msg.editEmbed {
                    embed(
                        SkipMessages.VOTE_SKIPPED,
                        Pair("{title}", track.title),
                        Pair("{author}", track.author)
                    )
                }
                    .setComponents()
                    .queue()

                LogUtilsKt(guild).sendLog(
                    LogType.TRACK_SKIP, SkipMessages.VOTE_SKIPPED_LOG,
                    Pair("{title}", track.title),
                    Pair("{author}", track.author)
                )
            }
    }

    fun clearVoteSkipInfo(guild: Guild) {
        val voteSkipManagerKt = RobertifyAudioManager[guild].voteSkipManager
        voteSkipManagerKt.clearVoters()
        voteSkipManagerKt.startedBy = null
        voteSkipManagerKt.isVoteSkipActive = false
        voteSkipManagerKt.voteSkipCount = 0

        val message = voteSkipManagerKt.voteSkipMessage
        if (message != null) {
            guild.getTextChannelById(message.first)
                ?.retrieveMessageById(message.second)
                ?.queue { msg ->
                    msg.editEmbed { embed(SkipMessages.SKIPPED) }
                        .queue { voteSkipManagerKt.voteSkipMessage = null }
                }
        }
    }

    private fun incrementVoteSkip(guild: Guild) {
        val voteSkipManager = RobertifyAudioManager[guild].voteSkipManager
        check(voteSkipManager.isVoteSkipActive) { "Can't increment vote skips since ${guild.name} (${guild.id}) doesn't have any active vote skips!" }
        voteSkipManager.voteSkipCount++
    }

    private fun decrementVoteSkip(guild: Guild) {
        val voteSkipManager = RobertifyAudioManager[guild].voteSkipManager
        check(voteSkipManager.isVoteSkipActive) { "Can't increment vote skips since ${guild.name} (${guild.id}) doesn't have any active vote skips!" }
        voteSkipManager.voteSkipCount--
    }

    private fun getNeededVotes(guild: Guild): Int {
        val size = guild.selfMember.voiceState!!.channel!!.members.size
        return ceil(size * (50 / 100.0)).toInt()
    }

    override suspend fun onButtonInteraction(event: ButtonInteractionEvent) {
        if (event.button.id?.startsWith("voteskip:") != true)
            return

        val split = event.button.id!!.split(":")
        val voteType = split[1]
        val guildId = split[2]

        val guild = event.guild!!
        if (guild.id != guildId)
            return

        val selfVoiceState = guild.selfMember.voiceState!!
        val memberVoiceState = event.member!!.voiceState!!

        if (!selfVoiceState.inAudioChannel()) {
            event.replyEmbed {
                embed(GeneralMessages.BUTTON_NO_LONGER_VALID)
            }.setEphemeral(true)
                .queue()
            return
        }

        if (!memberVoiceState.inAudioChannel()) {
            event.replyEmbed {
                embed(GeneralMessages.SAME_VOICE_CHANNEL_BUTTON)
            }.setEphemeral(true)
                .queue()
            return
        }

        if (selfVoiceState.channel!!.id != memberVoiceState.channel!!.id) {
            event.replyEmbed {
                embed(GeneralMessages.SAME_VOICE_CHANNEL_BUTTON)
            }.setEphemeral(true)
                .queue()
            return
        }

        val user = memberVoiceState.member.user
        val voteSkipManager = RobertifyAudioManager[guild].voteSkipManager
        when (voteType) {
            "upvote" -> {
                if (voteSkipManager.userAlreadyVoted(user.idLong)) {
                    voteSkipManager.removeVoter(user.idLong)
                    voteSkipManager.voteSkipCount--

                    event.replyEmbed {
                        embed(SkipMessages.SKIP_VOTE_REMOVED)
                    }
                        .setEphemeral(true)
                        .queue()
                    updateVoteSkipMessage(guild)
                } else {
                    if (GeneralUtils.hasPerms(guild, event.member, RobertifyPermission.ROBERTIFY_DJ)) {
                        doVoteSkip(guild)
                        event.replyEmbed {
                            embed(SkipMessages.DJ_SKIPPED)
                        }.setEphemeral(true)
                            .queue()
                        return
                    }

                    voteSkipManager.addVoter(user.idLong)
                    voteSkipManager.voteSkipCount++
                    event.replyEmbed {
                        embed(SkipMessages.SKIP_VOTE_ADDED)
                    }
                        .setEphemeral(true)
                        .queue()
                    updateVoteSkipMessage(guild)
                }
            }

            "cancel" -> {
                if (voteSkipManager.startedBy != user.idLong) {
                    event.replyEmbed {
                        embed(GeneralMessages.NO_PERMS_BUTTON)
                    }.setEphemeral(true)
                        .queue()
                    return
                }

                val message = voteSkipManager.voteSkipMessage!!
                guild.getTextChannelById(message.first)
                    ?.retrieveMessageById(message.second)
                    ?.submit()
                    ?.whenComplete { msg, err ->
                        if (err != null) {
                            logger.error("Unexpected error", err)
                            return@whenComplete
                        }

                        val track = RobertifyAudioManager[guild]
                            .player
                            .playingTrack!!

                        voteSkipManager.voteSkipMessage = null
                        clearVoteSkipInfo(guild)
                        msg.editEmbed {
                            embed(
                                SkipMessages.VOTE_SKIP_CANCELLED,
                                Pair("{title}", track.title),
                                Pair("{author}", track.author)
                            )
                        }.queue()
                    }
            }
        }
    }

    private fun updateVoteSkipMessage(guild: Guild) {
        val voteSkipManager = RobertifyAudioManager[guild].voteSkipManager
        if (!voteSkipManager.isVoteSkipActive)
            return

        val message = voteSkipManager.voteSkipMessage!!
        guild.getTextChannelById(message.first)
            ?.retrieveMessageById(message.second)
            ?.submit()
            ?.thenCompose { msg ->
                val neededvotes = getNeededVotes(guild)
                msg.editEmbed {
                    embed(
                        SkipMessages.VOTE_SKIP_STARTED_EMBED,
                        Pair(
                            "{user}",
                            GeneralUtils.toMention(
                                guild,
                                voteSkipManager.startedBy!!,
                                GeneralUtils.Mentioner.USER
                            )
                        ),
                        Pair("{neededVotes}", neededvotes.toString())
                    )
                }
                    .submit()
            }
            ?.whenComplete { _, err ->
                if (err != null) {
                    logger.error("Unexepected error", err)
                    return@whenComplete
                }
            }
    }

    override val help: String
        get() = "Skips a track or skip to a specific song in the queue"
}